第一章 概述/lidar_localization/third_part/GeographicLib-1.52/man/Planimeter.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.42)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PLANIMETER 1"
.TH PLANIMETER 1 "2021-06-21" "GeographicLib 1.52" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
Planimeter \-\- compute the area of geodesic polygons
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBPlanimeter\fR [ \fB\-r\fR ] [ \fB\-s\fR ] [ \fB\-l\fR ] [ \fB\-e\fR \fIa\fR \fIf\fR ]
[ \fB\-w\fR ] [ \fB\-p\fR \fIprec\fR ] [ \fB\-G\fR | \fB\-E\fR | \fB\-Q\fR | \fB\-R\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Measure the area of a geodesic polygon.  Reads polygon vertices from
standard input, one per line.  Vertices may be given as latitude and
longitude, \s-1UTM/UPS,\s0 or \s-1MGRS\s0 coordinates, interpreted in the same way as
\&\fBGeoConvert\fR\|(1).  (\s-1MGRS\s0 coordinates signify the center of the
corresponding \s-1MGRS\s0 square.)  The end of input, a blank line, or a line
which can't be interpreted as a vertex signals the end of one polygon
and the start of the next.  For each polygon print a summary line with
the number of points, the perimeter (in meters), and the area (in
meters^2).
.PP
The edges of the polygon are given by the \fIshortest\fR geodesic between
consecutive vertices.  In certain cases, there may be two or many such
shortest geodesics, and in that case, the polygon is not uniquely
specified by its vertices.  This only happens with very long edges (for
the \s-1WGS84\s0 ellipsoid, any edge shorter than 19970 km is uniquely
specified by its end points).  In such cases, insert an additional
vertex near the middle of the long edge to define the boundary of the
polygon.
.PP
By default, polygons traversed in a counter-clockwise direction return a
positive area and those traversed in a clockwise direction return a
negative area.  This sign convention is reversed if the \fB\-r\fR option is
given.
.PP
Of course, encircling an area in the clockwise direction is equivalent
to encircling the rest of the ellipsoid in the counter-clockwise
direction.  The default interpretation used by \fBPlanimeter\fR is the one
that results in a smaller magnitude of area; i.e., the magnitude of the
area is less than or equal to one half the total area of the ellipsoid.
If the \fB\-s\fR option is given, then the interpretation used is the one
that results in a positive area; i.e., the area is positive and less
than the total area of the ellipsoid.
.PP
Arbitrarily complex polygons are allowed.  In the case of
self-intersecting polygons the area is accumulated \*(L"algebraically\*(R",
e.g., the areas of the 2 loops in a figure\-8 polygon will partially
cancel.  Polygons may include one or both poles.  There is no need to
close the polygon.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-r\fR" 4
.IX Item "-r"
toggle whether counter-clockwise traversal of the polygon returns a
positive (the default) or negative result.
.IP "\fB\-s\fR" 4
.IX Item "-s"
toggle whether to return a signed result (the default) or not.
.IP "\fB\-l\fR" 4
.IX Item "-l"
toggle whether the vertices represent a polygon (the default) or a
polyline.  For a polyline, the number of points and the length of the
path joining them is returned; the path is not closed and the area is
not reported.
.IP "\fB\-e\fR \fIa\fR \fIf\fR" 4
.IX Item "-e a f"
specify the ellipsoid via the equatorial radius, \fIa\fR and
the flattening, \fIf\fR.  Setting \fIf\fR = 0 results in a sphere.  Specify
\&\fIf\fR < 0 for a prolate ellipsoid.  A simple fraction, e.g., 1/297,
is allowed for \fIf\fR.  By default, the \s-1WGS84\s0 ellipsoid is used, \fIa\fR =
6378137 m, \fIf\fR = 1/298.257223563.  If entering vertices as \s-1UTM/UPS\s0 or
\&\s-1MGRS\s0 coordinates, use the default ellipsoid, since the conversion of
these coordinates to latitude and longitude always uses the \s-1WGS84\s0
parameters.
.IP "\fB\-w\fR" 4
.IX Item "-w"
toggle the longitude first flag (it starts off); if the flag is on, then
when reading geographic coordinates, longitude precedes latitude (this
can be overridden by a hemisphere designator, \fIN\fR, \fIS\fR, \fIE\fR, \fIW\fR).
.IP "\fB\-p\fR \fIprec\fR" 4
.IX Item "-p prec"
set the output precision to \fIprec\fR (default 6); the perimeter is given
(in meters) with \fIprec\fR digits after the decimal point; the area is
given (in meters^2) with (\fIprec\fR \- 5) digits after the decimal point.
.IP "\fB\-G\fR" 4
.IX Item "-G"
use the series formulation for the geodesics.  This is the default
option and is recommended for terrestrial applications.  This option,
\&\fB\-G\fR, and the following three options, \fB\-E\fR, \fB\-Q\fR, and \fB\-R\fR, are
mutually exclusive.
.IP "\fB\-E\fR" 4
.IX Item "-E"
use \*(L"exact\*(R" algorithms (based on elliptic integrals) for the geodesic
calculations.  These are more accurate than the (default) series
expansions for |\fIf\fR| > 0.02.  (But note that the implementation of
areas in GeodesicExact uses a high order series and this is only
accurate for modest flattenings.)
.IP "\fB\-Q\fR" 4
.IX Item "-Q"
perform the calculation on the authalic sphere.  The area calculation is
accurate even if the flattening is large, \fIprovided\fR the edges are
sufficiently short.  The perimeter calculation is not accurate.
.IP "\fB\-R\fR" 4
.IX Item "-R"
The lines joining the vertices are rhumb lines instead of geodesics.
.IP "\fB\-\-comment\-delimiter\fR \fIcommentdelim\fR" 4
.IX Item "--comment-delimiter commentdelim"
set the comment delimiter to \fIcommentdelim\fR (e.g., \*(L"#\*(R" or \*(L"//\*(R").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing.  For a given polygon, the last such string found will be
appended to the output line (separated by a space).
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
print version and exit.
.IP "\fB\-h\fR" 4
.IX Item "-h"
print usage and exit.
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR \fIinfile\fR" 4
.IX Item "--input-file infile"
read input from the file \fIinfile\fR instead of from standard input; a file
name of \*(L"\-\*(R" stands for standard input.
.IP "\fB\-\-input\-string\fR \fIinstring\fR" 4
.IX Item "--input-string instring"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR \fIlinesep\fR" 4
.IX Item "--line-separator linesep"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR \fIoutfile\fR" 4
.IX Item "--output-file outfile"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of \*(L"\-\*(R" stands for standard output.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Example (the area of the 100km \s-1MGRS\s0 square 18SWK)
.PP
.Vb 7
\&   Planimeter <<EOF
\&   18n 500000 4400000
\&   18n 600000 4400000
\&   18n 600000 4500000
\&   18n 500000 4500000
\&   EOF
\&   => 4 400139.53295860 10007388597.1913
.Ve
.PP
The following code takes the output from gdalinfo and reports the area
covered by the data (assuming the edges of the image are geodesics).
.PP
.Vb 10
\&   #! /bin/sh
\&   egrep \*(Aq^((Upper|Lower) (Left|Right)|Center) \*(Aq |
\&   sed \-e \*(Aqs/d /d/g\*(Aq \-e "s/\*(Aq /\*(Aq/g" | tr \-s \*(Aq(),\er\et\*(Aq \*(Aq \*(Aq | awk \*(Aq{
\&       if ($1 $2 == "UpperLeft")
\&           ul = $6 " " $5;
\&       else if ($1 $2 == "LowerLeft")
\&           ll = $6 " " $5;
\&       else if ($1 $2 == "UpperRight")
\&           ur = $6 " " $5;
\&       else if ($1 $2 == "LowerRight")
\&           lr = $6 " " $5;
\&       else if ($1 == "Center") {
\&           printf "%s\en%s\en%s\en%s\en\en", ul, ll, lr, ur;
\&           ul = ll = ur = lr = "";
\&       }
\&   }
\&   \*(Aq | Planimeter | cut \-f3 \-d\*(Aq \*(Aq
.Ve
.SH "ACCURACY"
.IX Header "ACCURACY"
Using the \fB\-G\fR option (the default), the accuracy was estimated by
computing the error in the area for 10^7 approximately regular
polygons on the \s-1WGS84\s0 ellipsoid.  The centers and the orientations of
the polygons were uniformly distributed, the number of vertices was
log-uniformly distributed in [3, 300], and the center to vertex
distance log-uniformly distributed in [0.1 m, 9000 km].
.PP
The maximum error in the perimeter was 200 nm, and the maximum error
in the area was
.PP
.Vb 4
\&   0.0013 m^2 for perimeter < 10 km
\&   0.0070 m^2 for perimeter < 100 km
\&   0.070 m^2 for perimeter < 1000 km
\&   0.11 m^2 for all perimeters
.Ve
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBGeoConvert\fR\|(1), \fBGeodSolve\fR\|(1).
.PP
An online version of this utility is availbable at
<https://geographiclib.sourceforge.io/cgi\-bin/Planimeter>.
.PP
The algorithm for the area of geodesic polygon is
given in Section 6 of C. F. F. Karney, \fIAlgorithms for geodesics\fR,
J. Geodesy 87, 43\-55 (2013);
\&\s-1DOI\s0 <https://doi.org/10.1007/s00190\-012\-0578\-z>;
addenda: <https://geographiclib.sourceforge.io/geod\-addenda.html>.
.SH "AUTHOR"
.IX Header "AUTHOR"
\&\fBPlanimeter\fR was written by Charles Karney.
.SH "HISTORY"
.IX Header "HISTORY"
\&\fBPlanimeter\fR was added to GeographicLib,
<https://geographiclib.sourceforge.io>, in version 1.4.


第一章 概述/lidar_localization/third_part/GeographicLib-1.52/tools/Planimeter.cpp
/**
 * \file Planimeter.cpp
 * \brief Command line utility for measuring the area of geodesic polygons
 *
 * Copyright (c) Charles Karney (2010-2020) <charles@karney.com> and licensed
 * under the MIT/X11 License.  For more information, see
 * https://geographiclib.sourceforge.io/
 *
 * See the <a href="Planimeter.1.html">man page</a> for usage information.
 **********************************************************************/

#include <iostream>
#include <string>
#include <sstream>
#include <fstream>
#include <GeographicLib/PolygonArea.hpp>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>
#include <GeographicLib/GeoCoords.hpp>
#include <GeographicLib/Ellipsoid.hpp>

#if defined(_MSC_VER)
// Squelch warnings about constant conditional expressions
#  pragma warning (disable: 4127)
#endif

#include "Planimeter.usage"

int main(int argc, const char* const argv[]) {
  try {
    using namespace GeographicLib;
    typedef Math::real real;
    Utility::set_digits();
    enum { GEODESIC, EXACT, AUTHALIC, RHUMB };
    real
      a = Constants::WGS84_a(),
      f = Constants::WGS84_f();
    bool reverse = false, sign = true, polyline = false, longfirst = false;
    int linetype = GEODESIC;
    int prec = 6;
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';';

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-r")
        reverse = !reverse;
      else if (arg == "-s")
        sign = !sign;
      else if (arg == "-l")
        polyline = !polyline;
      else if (arg == "-e") {
        if (m + 2 >= argc) return usage(1, true);
        try {
          a = Utility::val<real>(std::string(argv[m + 1]));
          f = Utility::fract<real>(std::string(argv[m + 2]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -e: " << e.what() << "\n";
          return 1;
        }
        m += 2;
      } else if (arg == "-w")
        longfirst = !longfirst;
      else if (arg == "-p") {
        if (++m == argc) return usage(1, true);
        try {
          prec = Utility::val<int>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "-G")
        linetype = GEODESIC;
      else if (arg == "-E")
        linetype = EXACT;
      else if (arg == "-Q")
        linetype = AUTHALIC;
      else if (arg == "-R")
        linetype = RHUMB;
      else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout << argv[0] << ": GeographicLib version "
                  << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else
        return usage(!(arg == "-h" || arg == "--help"), arg != "--help");
    }

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    const Ellipsoid ellip(a, f);
    if (linetype == AUTHALIC) {
      using std::sqrt;
      a = sqrt(ellip.Area() / (4 * Math::pi()));
      f = 0;
    }
    const Geodesic geod(a, f);
    const GeodesicExact geode(a, f);
    const Rhumb rhumb(a, f);
    PolygonArea poly(geod, polyline);
    PolygonAreaExact polye(geode, polyline);
    PolygonAreaRhumb polyr(rhumb, polyline);
    GeoCoords p;

    // Max precision = 10: 0.1 nm in distance, 10^-15 deg (= 0.11 nm),
    // 10^-11 sec (= 0.3 nm).
    prec = std::min(10 + Math::extra_digits(), std::max(0, prec));
    std::string s, eol("\n");
    real perimeter, area;
    unsigned num;
    while (std::getline(*input, s)) {
      if (!cdelim.empty()) {
        std::string::size_type m = s.find(cdelim);
        if (m != std::string::npos) {
          eol = " " + s.substr(m) + "\n";
          s = s.substr(0, m);
        }
      }
      bool endpoly = s.empty();
      if (!endpoly) {
        try {
          using std::isnan;
          p.Reset(s, true, longfirst);
          if (isnan(p.Latitude()) || isnan(p.Longitude()))
            endpoly = true;
        }
        catch (const GeographicErr&) {
          endpoly = true;
        }
      }
      if (endpoly) {
        num =
          linetype == EXACT ? polye.Compute(reverse, sign, perimeter, area) :
          linetype == RHUMB ? polyr.Compute(reverse, sign, perimeter, area) :
          poly.Compute(reverse, sign, perimeter, area); // geodesic + authalic
        if (num > 0) {
          *output << num << " " << Utility::str(perimeter, prec);
          if (!polyline) {
            *output << " " << Utility::str(area, std::max(0, prec - 5));
          }
          *output << eol;
        }
        linetype == EXACT ? polye.Clear() :
          linetype == RHUMB ? polyr.Clear() : poly.Clear();
        eol = "\n";
      } else {
        linetype == EXACT ? polye.AddPoint(p.Latitude(), p.Longitude()) :
          linetype == RHUMB ? polyr.AddPoint(p.Latitude(), p.Longitude()) :
          poly.AddPoint(linetype == AUTHALIC ?
                        ellip.AuthalicLatitude(p.Latitude()) :
                        p.Latitude(),
                        p.Longitude());
      }
    }
    num =
      linetype == EXACT ? polye.Compute(reverse, sign, perimeter, area) :
      linetype == RHUMB ? polyr.Compute(reverse, sign, perimeter, area) :
      poly.Compute(reverse, sign, perimeter, area);
    if (num > 0) {
      *output << num << " " << Utility::str(perimeter, prec);
      if (!polyline) {
        *output << " " << Utility::str(area, std::max(0, prec - 5));
      }
      *output << eol;
    }
    linetype == EXACT ? polye.Clear() :
      linetype == RHUMB ? polyr.Clear() : poly.Clear();
    eol = "\n";
    return 0;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
